## Worksheet 01b: Writing functions & tests
_**Leader:** Icíar Fernández **Reviewer:** Victor Yuan **ASDA Assist:** David Kepplinger_

_Version 1_

### Attributions

Content from this worksheet is largely based on:

+ Jenny Bryan's STAT545 [Write your own functions](https://stat545.com/functions-part1.html)
    + Some prose is taken verbatim!
+ [R Exercises](https://www.r-exercises.com/start-here-to-learn-r/)

### Introduction

This is the corresponding worksheet for Class 1 (October 27th, 2020) & Class 2 (October 29th, 2020).

There are 8 questions on this worksheet. To get 100%, you must get 50% of questions of the worksheet correct - 0.5 * 8 = **4 questions**. 

Some notes:
+ Remember to pay attention to the variable name you store your answer in, or else it will not be autograded correctly.
+ To ensure everything works properly, remember to run all code cells, not just the ones with your answer.

If you want to use packages which are not yet installed, you can use the code cell below to install them. You may not have the packages `palmerpenguins` and `testthat` installed.

In [ ]:
# Install additional packages, e.g.

# install.packages("palmerpenguins")
# install.packages("lubridate")
# install.packages("gapminder")
# install.packages("tidyverse")
# install.packages("testthat")
# your code here
fail() # No Answer - remove if you provide an answer

Use the following code cell to load any additional packages you want to use for this worksheet.

In [ ]:
# Load packages, e.g.
# library(devtools)
# your code here
fail() # No Answer - remove if you provide an answer

Run the code cell below to load the packages.

In [1]:
suppressPackageStartupMessages(library(palmerpenguins))
suppressPackageStartupMessages(library(lubridate))
suppressPackageStartupMessages(library(gapminder))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(testthat))

### Class 1: Writing functions

**QUESTION 1:**
Create a function that allows you to compute the max minus min of the Adelie penguins body mass. A code snippet to calculate max minus min *without a function* is shown to help you - essentially, your task is to turn this code snippet into a function. Be sure to make all three objects shown below.

```r
# put into practice your knowledge of dplyr to subset the penguins dataset to only those from the Adelie species
adelie <- penguins %>% 
    filter(FILL_THIS_IN == FILL_THIS_IN) %>% 
    drop_na()

# write your function here
max_minus_min <- function (x) FILL_THIS_IN - FILL_THIS_IN

# apply your function to the Adelie penguins body mass
answer1.0 <- max_minus_min(FILL_THIS_IN)
```

In [7]:
# put into practice your knowledge of dplyr to subset the penguins dataset to only those from the Adelie species
adelie <- penguins %>% 
    filter(species == "Adelie") %>% 
    drop_na()

# write your function here
max_minus_min <- function (x) max(x) - min(x)

# apply your function to the Adelie penguins body mass
answer1.0 <- max_minus_min(adelie$body_mass_g)
head(adelie)
cat("Your final answer:")
answer1.0

species island    bill_length_mm bill_depth_mm flipper_length_mm body_mass_g
1 Adelie  Torgersen 39.1           18.7          181               3750       
2 Adelie  Torgersen 39.5           17.4          186               3800       
3 Adelie  Torgersen 40.3           18.0          195               3250       
4 Adelie  Torgersen 36.7           19.3          193               3450       
5 Adelie  Torgersen 39.3           20.6          190               3650       
6 Adelie  Torgersen 38.9           17.8          181               3625       
  sex    year
1 male   2007
2 female 2007
3 female 2007
4 female 2007
5 male   2007
6 female 2007

Your final answer:

[1] 1925

In [8]:
test_that("Question 1", {
    expect_known_hash(sort(round(adelie$body_mass_g, 2)), '4c17a7083ab7d3e61b770cfd1ea2515d')
    expect_known_hash(round(answer1.0, 3), '112052893c8bd4663fea8754262dfb9e')
})
cat("Success!")

Success!

**QUESTION 2:** Test your function on the life expectancy variable of the `gapminder` dataset and assign the returned value to R object `answer2.0`. Does it work?

In [9]:
answer2.0 <- max_minus_min(gapminder $lifeExp)

print(answer2.0)

[1] 59.004


In [10]:
test_that("Question 2", {
    expect_known_hash(round(answer2.0, 3), '4ca07c689295be575d7b41d1c7d8c61f')
})
cat("Success!")

Success!

**QUESTION 3:** Now, run the following code to test your function. As you can see, the function doesn't work on categorical variables, tibbles or strings - which we expect, because it wouldn't make sense to compute the max minus min in any of those arguments. 

In [ ]:
#max_minus_min(adelie)
#max_minus_min(penguins$species)
#max_minus_min("stat545 is great")
# your code here
fail() # No Answer - remove if you provide an answer

However, R will do *anything* to try and make sense of your function... and that is not always a good thing. If you run the code below, you will see that the function is giving us an output for arguments that do not make any sense.

In [11]:
max_minus_min(gapminder[c('lifeExp', 'gdpPercap', 'pop')])
max_minus_min(c(TRUE, FALSE, TRUE))

[1] 1318683072

[1] 1

To avoid this from happening, rewrite the `max_minus_min` function to include a `stopifnot()`. Think of what should be the argument of `stopifnot()` to solve this issue - in other words, what should be the object class that the function `max_minus_min` should accept (e.g. a dataframe, a character, a number...)?

```r
answer3.0 <- function(x) {
  stopifnot(FILL_THIS_IN) 
  max(x) - min(x)
}
```

In [16]:
answer3.0 <- function(x) {
  stopifnot(is.numeric(x)) 
  max(x) - min(x)
}
answer3.0

function(x) {
  stopifnot(is.numeric(x)) 
  max(x) - min(x)
}

In [17]:
test_that("Question 3", {
    expect_known_hash(mode(answer3.0), 'ecce171f95c8c2466c6516b40beca466')
    expect_known_hash(length(formals(answer3.0)), '4b5630ee914e848e8d07221556b0a2fb')
    expect_error(answer3.0('a'), 'is.numeric')
    expect_error(answer3.0(c(TRUE, FALSE)), 'is.numeric')
    expect_error(answer3.0(list(1:5)), 'is.numeric')
    expect_known_hash(answer3.0(1:5), '234a2a5581872457b9fe1187d1616b13')
})
cat("Success!")

Success!

**QUESTION 4:** In the following chunk, I created a function to convert fahrenheit to celsius. When I test it with arguments that theoretically shouldn't give an output, it appears to have the same problem as our original `max_minus_min()` function. Rewrite the `fahrenheit_to_celsius` function using `if()` and `stop()` instead of `stopifnot()`. This allows you to write your own (more informative) error message.

*Hint:* Remember that you are trying to stop the function from working **if the argument is not numeric.**

In [ ]:
# here I write the function
fahrenheit_to_celsius <- function(temp_F) {
  temp_C <- (temp_F - 32) * 5 / 9
  return(temp_C)
}

In [ ]:
# here I test it on something that theoretically should not work.. but does
fahrenheit_to_celsius(c(TRUE, FALSE, FALSE, TRUE))

```r
answer4.0 <- function(temp_F) {
  if(!FILL_THIS_IN(FILL_THIS_IN)) {
    stop('I am so sorry, but this function only works for numeric input!\n',
         'You have provided an object of class: ', class(FILL_THIS_IN)[1])
  }
  temp_C <- (temp_F - 32) * 5 / 9
  return(FILL_THIS_IN)
}
```

In [18]:
answer4.0 <- function(temp_F) {
  if(!is.numeric(temp_F)) {
    stop('I am so sorry, but this function only works for numeric input!\n',
         'You have provided an object of class: ', class(temp_F)[1])
  }
  temp_C <- (temp_F - 32) * 5 / 9
  return(temp_C)
}

In [19]:
test_that("Question 4", {
    expect_known_hash(mode(answer4.0), 'ecce171f95c8c2466c6516b40beca466')
    expect_known_hash(length(formals(answer4.0)), '4b5630ee914e848e8d07221556b0a2fb')
    expect_error(answer4.0(c('4', '20', '12')))
    expect_error(answer4.0(c(TRUE, FALSE)))
    expect_error(answer4.0(list(1:5)))
    expect_known_hash(answer4.0(c(350, 425, 550)), 'ebd71f9f1506f215d50b4d64546b8f7e')
    expect_known_hash(answer4.0(1:5), '1faaa13b49412d6a48b0d25c0011f38f')
})
cat("Success!")

Success!

Let's try testing your new function with a non-numeric argument below to see what happens! 

In [ ]:
#answer4.0(penguins)
# your code here
fail() # No Answer - remove if you provide an answer

**QUESTION 5.0:** Write a function that takes 2 arguments (name them `x` and `y`), raises the first argument to the power of the second one, and prints the result with a message that indicates what the output of the function is. Use `stopifnot()` so that the function *only* takes numeric arguments.

```r
# function to print x raised to the power y
answer5.0 <- function(FILL_THIS_IN, FILL_THIS_IN) {
    stopifnot(FILL_THIS_IN && FILL_THIS_IN)
    result <- FILL_THIS_IN
    cat(FILL_THIS_IN, "raised to the power", FILL_THIS_IN, "is", FILL_THIS_IN)
}
```

In [2]:
# function to print x raised to the power y
answer5.0 <- function(x, y) {
    stopifnot(is.numeric(x) && is.numeric(y))
    result <- x^y
    cat(result, "raised to the power", x, "is", y)
}
answer5.0(3, 4)

81 raised to the power 3 is 4

In [3]:
test_that("Question 5", {
    expect_known_hash(mode(answer5.0), 'ecce171f95c8c2466c6516b40beca466')
    expect_known_hash(length(formals(answer5.0)), 'c01f179e4b57ab8bd9de309e6d576c48')
    expect_error(answer5.0(3, c('4', '20', '12')), 'is.numeric')
    expect_error(answer5.0(c(TRUE, FALSE), 2), 'is.numeric')
    expect_error(answer5.0(list(1:5), 1:3), 'is.numeric')
    expect_output(answer5.0(1/5, 4), '0\\.0016')
    expect_output(answer5.0(2, 2:5), '4 8 16 32')
})
cat("Success!")

Success!

### Testing functions

`testthat` is a collection of functions developed by Hadley Wickham that makes unit testing easy for developers. You can read more about the structure of tests [here](https://r-pkgs.org/tests.html#test-structure). In a nutshell, tests are organized hierarchically: **expectations** are grouped into **tests**. Functions that start with `expect_` describe the expected result of a computation (e.g. Does it have the right class?) - these are expectations. Tests are created with `test_that()` and they group together multiple *expectations* to test the output of a function (at their simplest level). 

**QUESTION 6:** Expectation functions have two arguments, the first is the actual result, and the second is what you expect. Use `expect_equal()` to check that manually calculating the max minus min of the bill length of all penguins in the `penguins` dataset yields the same result as using a function.

Run the following chunk of code first.

In [4]:
x <- max(penguins$bill_length_mm, na.rm = TRUE) 
y <- min(penguins$bill_length_mm, na.rm = TRUE) 
x - y

[1] 27.5

I have written a slightly tweaked version of the `max_minus_min()` function created in **Q1** that allows the user to control the behaviours around NAs. This is important - without removing NAs, if you try running `max_minus_min()` with `penguins$bill_length_mm` as an argument, the output will be NA.

In [5]:
# new function
max_minus_min2 <- function(x, na.rm = TRUE) {
  if(!is.numeric(x)) {
    stop('I am so sorry, but this function only works for numeric input!\n',
         'You have provided an object of class: ', class(x)[1])
  }
  max(x, na.rm = na.rm) - min(x, na.rm = na.rm)
}

Now, check that the output of this function when calculating the max minus min of the bill length across all penguins in the `penguins` dataset is the same as calculating it "manually" as we did above - in other words, that the function output **equals** 27.5.

__NOTE__: We're _only_ getting you to store the output of `expect_equal()` in variable `answer6.0` so that we can run the autograder! Otherwise, you'd only ever run it on its own, without assigning it to anything.

_Psst... take a look at the test cells in these worksheets. Notice a similarity?_

```r
answer6.0 <- expect_equal(FILL_THIS_IN, FILL_THIS_IN(FILL_THIS_IN$FILL_THIS_IN, na.rm = TRUE))
```

In [6]:
answer6.0 <- expect_equal(x-y, max_minus_min2(penguins$bill_length_mm, na.rm = TRUE))
answer6.0

[1] 27.5

In [7]:
test_that("Question 6", {
    expect_known_hash(answer6.0, 'b342fe04b00f00610a95dd8ebcc5967c')
})
cat("Success!")

Success!

There are other `expect_` functions such as `expect_identical()`, `expect_match()` or `expect_output()` that you can look into further [here](https://r-pkgs.org/tests.html#test-structure). 

**QUESTION 7:** Let's try combining a different expectation function and `test_that()` with a very easy example. We start by writing a very simple function that returns the string "Hello world, my name is" + your name.

In [8]:
hello_world <- function(x) {
    stopifnot(is.character(x)) # we want x to be a character
    paste('Hello world, my name is', x)
}

In the next code cell, we'll run `hello_world()` with "Julie Payette" as an argument, just to see the output!

In [9]:
hello_world("Julie Payette")

[1] "Hello world, my name is Julie Payette"

Great! Now, let's simply test that the output of `hello_world()` with _your_ name as an argument matches the character vector **'Hello world, my name is'**. I have added an example of how `expect_match()` works to help you write your test.

```
eggplants <- "Eggplants are purple"
expect_match(eggplants, "Eggplants") # works
expect_match(eggplants, "purple") # works

# your turn
answer7.0 <- test_that("returns hello world + your name string", {
    expect_match(FILL_THIS_IN(FILL_THIS_IN), FILL_THIS_IN)
})
```

In [11]:
# your turn
answer7.0 <- test_that("returns hello world + your name string", {
    expect_match(hello_world("Julie Payette"), "Hello world, my name is")
})

In [12]:
test_that("Question 7", {
    expect_known_hash(answer7.0, 'bb73ad91bcb7e948250d465016f7b')
})
cat("Success!")

Success!

**QUESTION 8:** Create a function called `m` that multiplies two numbers (arguments `x` and `y`). Create a test for function `m` with description "Testing multiplication function" and add a few scenarios to it:

+ Check if `m(2,3)` equals `6`
+ Check if `m(2, c(2,3))` equals `c(4,6)`
+ Check if `m(2, "3")` throws an error "non-numeric argument to binary operator"

```r
m <- function(FILL_THIS_IN) FILL_THIS_IN
answer8.0 <- test_that("Testing multiplication function", {
  expect_equal(FILL_THIS_IN)
  expect_equal(FILL_THIS_IN)
  expect_error(FILL_THIS_IN, FILL_THIS_IN)
})
```

In [17]:
m <- function(x,y) {
     if(!is.numeric(x) || !is.numeric(y)) {
        stop("non-numeric argument to binary operator")
         }
    x*y
}
answer8.0 <- test_that("Testing multiplication function", {
  expect_equal(m(2,3),6)
  expect_equal(m(2, c(2,3)),c(4,6))
  expect_error(m(2, "3"), "non-numeric argument to binary operator")
})

In [18]:
test_that("Question 8", {
    expect_known_hash(round(answer8.0, 2), '6717f2823d3202449301145073ab8')
})
cat("Success!")

Success!